lib/core/test/unit/Cardano/Wallet/Gen.hs

```diff
@@ -259,3 +270,47 @@ shrinkTxMetadataValue (TxMetaList xs) =
 shrinkTxMetadataValue (TxMetaNumber i) = TxMetaNumber <$> shrink i
 shrinkTxMetadataValue (TxMetaBytes b) = TxMetaBytes <$> shrinkByteString b
 shrinkTxMetadataValue (TxMetaText s) = TxMetaText <$> shrinkText s
+
+genNatural :: Gen Natural
+genNatural = arbitrarySizedNatural
+
+genScript :: [KeyHash] -> Gen (Script KeyHash)
+genScript keyHashes = scale (`div` 3) $ sized scriptTree
+    where
+        scriptTree 0 = oneof
+            [ RequireSignatureOf <$> elements keyHashes
+            , ActiveFromSlot <$> genNatural
+            , ActiveUntilSlot <$> genNatural
+            ]
+        scriptTree n = do
```

Comment:
Shouldn't the generator be more agnostic to the script underlying structure and simply generate truly arbitrary scripts. Some of them may contain invalid timelocks, which is fine. We can use quickcheck labels to measure coverage and make sure that we cover interesting scenarios. 

The risk of having a generator that is _too complex_ and too close to what the actual implementation expects is that, we may end up introducing bugs present in the implementation in the generator itself and, fail to catch wrong behaviors from the implementation. As far as I am concerned, the generator for a script could be `genericArbitrary` and would not need a lot of tweaking. This may generate scripts that can be considered "invalid" by our functions yet, such scripts are still plausible and should therefore be part of the generated set of inputs. 

If some properties only apply to wellformed scripts then it should be possible to discard wrong cases with a pre-condition `==>`

Reply:
thought some more about, experimented with things like : 
```
genScriptValidated :: [KeyHash] -> Gen (Script KeyHash)                                                                                                                                                        
genScriptValidated keyHashes =                                                                                                                                                                                 
    genScript keyHashes `suchThat` (isRight . validateScript RecommendedValidation)
```
which slowed generation considerably.
In the end I went with the simplest solution - maximum possible simplification

Reply:
Simple is good. It looks simpler now.
It's also possible to add some cover tests to ensure the generator makes some "interesting" valid test cases.

With any recursive type, I'm always a bit worried about the generator making things too big, or the shrinker being too slow... can you reassure me that this generator is going to be nice to us in future?